container/credentials.bzl

```diff
@@ -0,0 +1,118 @@
+load("//container:providers.bzl", "RegistryCredentialInfo")
+
+def write_docker_config_json_action(ctx, registry, credential, config_json = None):
+    """Create an action that writes a docker config file.
+
+    Args:
+      ctx: the rule ctx object
+      registry: the name of the registry to which the credential should be
+      associated
+      credential: a RegistryCredentialInfo struct
+      config_json: path to the file where the config.json should be written.
+      Defaults to %{credential.name}/.docker/config.json.
+
+    Return: The config_json file 
+    """
+    if credential.type != "basic":
+        fail("Unsupported docker credential type: %s" % credential.type)
+    if not config_json:
+        config_json = ctx.actions.declare_file("%s/.docker/config.json" % credential.name)
+    # NOTE: ctx.actions.write does not work here as we need the
+    # use_default_shell_env feature of the run_shell action.
+    ctx.actions.run_shell(
+        mnemonic = "WriteDockerConfig",
+        inputs = [],
+        outputs = [config_json],
+        command = """
+set -euo pipefail
+cat <<EOF > "%s"
+{
+    "auths": {
+        "%s": {
+            "username": "${%s}",
+            "password": "${%s}"
+        }
+    }
+}
+EOF
+        """ % (config_json.path, registry, credential.user_env, credential.pass_env),
+        use_default_shell_env = True,
+    )
+    return config_json
+
+
+def _registry_credential_impl(ctx):
+    """Implementation of the registry credential
+
+    Provides the RegistryCredentialInfo object from given rule attributes.
+    """
+    return [RegistryCredentialInfo(
+        name = ctx.label.name,
+        user_env = ctx.attr.user_env,
+        pass_env = ctx.attr.pass_env,
+        type = ctx.attr.type,
+    )]
+
+registry_credential = rule(
+    implementation = _registry_credential_impl,
+    attrs = {
+        "user_env": attr.string(
```

Comment:
if this was a repo rule, env variables to be read would be declared in a way that Bazel knows to regenerate outputs if values change.
example:
https://github.com/bazelbuild/bazel/blob/master/tools/cpp/cc_configure.bzl#L59

Reply:
See comment above.  It doesn't have to be a repository rule.